### PR TITLE
API: EA._can_hold_na -> EDtype.can_hold_na

### LINKED ISSUE #40574: API: EA._can_hold_na should be a dtype attribute, not EA
In practice in all the cases I'm aware of, _can_hold_na is always fixed at the EA subclass level, so we can immediately infer _can_hold_na from the dtype.  But in principle ATM an EA subclass could do

```
class WeirdEA(ExtensionArray):
      @property
      def _can_hold_na(self) -> bool:
           return some_func(self[0])
```

We should make _can_hold_na an attribute of `ExtensionDtype` to pre-empt this.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #40574
- [ ] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -158,6 +158,9 @@ class DatetimeLikeArrayMixin(OpsMixin, NDArrayBackedExtensionArray):
     _is_recognized_dtype: Callable[[DtypeObj], bool]
     _recognized_scalars: tuple[type, ...]
     _ndarray: np.ndarray
+    # Incompatible types in assignment (expression has type "bool", base class
+    # "ExtensionArray" defined the type as "Callable[[ExtensionArray], bool]")
+    _can_hold_na = True  # type: ignore[assignment]
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py:163]: if u make this a property (like u do below) then can avoid the typing issue

File: pandas/tests/extension/test_external_block.py [hunk 2]
```diff
@@ -14,9 +14,9 @@
 class CustomBlock(ExtensionBlock):
 
     _holder = np.ndarray
-    # error: Cannot override final attribute "_can_hold_na"
-    # (previously declared in base class "Block")
-    _can_hold_na = False  # type: ignore[misc]
+    # Incompatible types in assignment (expression has type "bool", base class
+    # "Block" defined the type as "Callable[[Block], bool]")
```
Comment (jreback) [inline pandas/tests/extension/test_external_block.py:18]: same comment here

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]
lgtm. @jorisvandenbossche if any comments

**Comment by jreback** [comment]
> pandas/tests/extension/test_external_block.py:18: error: Cannot override final attribute "_can_hold_na" (previously declared in base class "Block")  [misc]

actually this is failing precommit

**Comment by jbrockmendel** [comment]
pre-commit fixed

File: pandas/core/dtypes/base.py [hunk 3]
```diff
@@ -367,6 +367,13 @@ def _get_common_dtype(self, dtypes: list[DtypeObj]) -> DtypeObj | None:
         else:
             return None
 
+    @property
+    def can_hold_na(self) -> bool:
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/base.py:371]: This doesn't need to be public?

**Comment by jorisvandenbossche** [comment]
I know this is an existing property on the ExtensionArray, but do we actually support (or want to support) an extension array that cannot hold missing values? (I don't think we currently have one, so this is also not really tested?)

**Comment by jorisvandenbossche** [comment]
> I don't think we currently have one

Except for PandasArray, I assume (but since that's not actually stored in practice, that's a bit a special case anyhow)

**Comment by jbrockmendel** [comment]
I think `Sparse[inty]`?

**Comment by simonjayhawkins** [comment]
i've added blocker since I don't think we should change this during the rc. so we either get it in for 1.3.0rc or move to 1.4/2.0

**Comment by jbrockmendel** [comment]
is the public/private question (and a rebase) the only outstanding issue?  im happy to make this private

**Comment by jorisvandenbossche** [comment]
Also the question whether we should have this at all, IMO. And if we have that property, what does it mean? (what's the behaviour for operations like setitem or reindexing that can introduce missing values?)

**Comment by jbrockmendel** [comment]
> Also the question whether we should have this at all, IMO. And if we have that property, what does it mean? (what's the behaviour for operations like setitem or reindexing that can introduce missing values?)

I'm open to having this discussion, but I think it can be done separately from the "where should this attribute live" discussion given that it already exists on EA.

**Comment by jorisvandenbossche** [comment]
But if we are not yet sure if and how such property would be used, we can also leave it where it is for now? (or what problem is this PR fixing?)

If moving it, let's at least make it private and document it that way (as an internal property that no EA author should change, as that is undefined behaviour)

**Comment by jbrockmendel** [comment]
> (or what problem is this PR fixing?)

The corner cases I want to rule out are where can_hold_na is values-dependent and potentially non-constant, e.g. (xref #40574)

```
class WeirdEA(ExtensionArray):
      @property
      def _can_hold_na(self) -> bool:
           return some_func(self[0])
```

**Comment by jreback** [comment]
this lgtm, i agree this should be on the dtype.

**Comment by jreback** [comment]
@jorisvandenbossche ok here?

File: doc/source/whatsnew/v1.3.0.rst [hunk 4]
```diff
@@ -644,6 +644,7 @@ Other API changes
 - Accessing ``_constructor_expanddim`` on a :class:`DataFrame` and ``_constructor_sliced`` on a :class:`Series` now raise an ``AttributeError``. Previously a ``NotImplementedError`` was raised (:issue:`38782`)
 - Added new ``engine`` and ``**engine_kwargs`` parameters to :meth:`DataFrame.to_sql` to support other future "SQL engines". Currently we still only use ``SQLAlchemy`` under the hood, but more engines are planned to be supported such as ``turbodbc`` (:issue:`36893`)
 - Removed redundant ``freq`` from :class:`PeriodIndex` string representation (:issue:`41653`)
+- :attr:`ExtensionArray._can_hold_na` should now be implemented on the :class:`ExtensionDtype` subclass as ``_can_hold_na`` (:issue:`40574`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.3.0.rst:647]: I would remove this mention, if we are not yet clear on the intended semantics of it.

**Comment by jorisvandenbossche** [comment]
Yeah, I won't object moving it. Just a tiny comment on the whatsnew.

**Comment by jbrockmendel** [comment]
whatsnew removed + greenish

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel

**Pull request merged**